apollo-router/src/services/router_service.rs

```diff
@@ -339,61 +326,219 @@ impl RouterService {
-                                            GRAPHQL_JSON_RESPONSE_HEADER_VALUE,
-                                            MULTIPART_DEFER_CONTENT_TYPE
-                                        ))
-                                        .extension_code("INVALID_ACCEPT_HEADER")
-                                        .build(),
-                                )
-                                .status_code(StatusCode::NOT_ACCEPTABLE)
-                                .header(CONTENT_TYPE, APPLICATION_JSON.essence_str())
-                                .context(context)
-                                .build()
+                    results.push(router::Response::error_builder()
+                            .error(
+                                graphql::Error::builder()
+                                    .message(format!(
+                                        r#"'accept' header must be one of: \"*/*\", {:?}, {:?} or {:?}"#,
+                                        APPLICATION_JSON.essence_str(),
+                                        GRAPHQL_JSON_RESPONSE_HEADER_VALUE,
+                                        MULTIPART_DEFER_CONTENT_TYPE
+                                    ))
+                                    .extension_code("INVALID_ACCEPT_HEADER")
+                                    .build(),
+                            )
+                            .status_code(StatusCode::NOT_ACCEPTABLE)
+                            .header(CONTENT_TYPE, APPLICATION_JSON.essence_str())
+                            .context(context)
+                            .build()?);
                 }
             }
         }
+        Ok(())
+    }
+
+    async fn call_inner(&self, req: RouterRequest) -> Result<RouterResponse, BoxError> {
+        let context = req.context.clone();
+
+        let supergraph_requests = match self.translate_request(req).await {
+            Ok(requests) => requests,
+            Err(err) => {
+                ::tracing::error!(
+                    monotonic_counter.apollo_router_http_requests_total = 1u64,
+                    status = %err.status.as_u16(),
+                    error = %err.error,
+                );
+
+                return router::Response::error_builder()
+                    .error(
+                        graphql::Error::builder()
+                            .message(String::from("Invalid GraphQL request"))
+                            .extension_code(err.extension_code)
+                            .extension("details", err.extension_details)
+                            .build(),
+                    )
+                    .status_code(err.status)
+                    .header(CONTENT_TYPE, APPLICATION_JSON.essence_str())
+                    .context(context)
+                    .build();
+            }
+        };
+
+        let mut results = Vec::with_capacity(supergraph_requests.len());
+
+        for supergraph_request in supergraph_requests {
```

Comment:
Does this mean that each request is currently processed one at a time rather than concurrently?

Reply:
Yes, to preserve ordering. Otherwise, I'd be concerned about mutation side-effects and the difficulties of re-assembling the batch response out of order.

Reply:
Did the original user submitted PR do parallel execution? (Also worth checking the gateway)

Reply:
The original user submitted PR did concurrent execution. I don't think it should have because of the "mutation side-effects" point from above.

It might be possible to check if batch contains mutations and execute concurrently if it doesn't. That would be a viable followup issue for performance.

Reply:
resolution: batches should be executed concurrently.
done: [88c8326](https://github.com/apollographql/router/pull/3837/commits/88c83261b315afb6d309fdc087606c7ce292bf61)